### PR TITLE
Adding ability to add coreq, prereq and update courses

### PR DESCRIPTION
**Pull request by si-x-li**

### CODE REVIEW

File: backend/routes/courses.js [hunk 1]
```diff
@@ -149,4 +148,154 @@ router.get("/getAllCourses", async (req, res, next) => {
...
+ *          database connection handling
+ *
+ * @author: Steven Li
+ */
+router.post("/addCoreq", async (req, res, next) => {
+  let coreqJSON;
+  try {
+    coreqJSON = JSON.parse(req.body.coreq);
+  } catch (err) {
+    res.status(500).send("invalid format coreq");
```
Comment (KevenLYF) [inline backend/routes/courses.js]: Invalid user input should return 400

File: backend/logic/courses/courses.js [hunk 2]
```diff
@@ -145,9 +145,128 @@ var addCourseOfferings = async courseOfferings => {
...
+    "INSERT INTO course_prereq (course_code, prereq_course_code) VALUES (?, ?);";
+  try {
+    await connection.beginTransaction();
+    for (let course in prereq) {
+      for (let i = 0; i < prereq[course].length; i++) {
+        await connection.query(query, [course, prereq[course][i]]);
+      }
+    }
+    await connection.commit();
+    connection.release();
```
Comment (KevenLYF) [inline backend/logic/courses/courses.js]: you can do `connection.release()` in finally clause, check my PR #42

File: backend/routes/courses.js [hunk 3]
```diff
@@ -149,4 +148,154 @@ router.get("/getAllCourses", async (req, res, next) => {
...
+    return;
```
Comment (KevenLYF) [inline backend/routes/courses.js]: just do `return res.status(400).send("invalid format coreq");`

File: backend/routes/courses.js [hunk 4]
```diff
@@ -149,4 +148,154 @@ router.get("/getAllCourses", async (req, res, next) => {
...
+router.post("/updateCourse", async (req, res, next) => {
+  let course = req.body.course;
+  let newTitle = req.body.new_title;
+  let newTagsJSON;
+  try {
+    newTagsJSON = JSON.parse(req.body.new_tags);
+  } catch (err) {
+    console.error(err);
+    res.status(500).send(err.message);
+    return;
```
Comment (KevenLYF) [inline backend/routes/courses.js]: same as before

**Comment by KevenLYF** [comment]
Check my comments, nice work you are on fire xd !

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (si-x-li) [reply backend/logic/courses/courses.js]: will finally even run if you're returning a value in the try block?

File: backend/routes/courses.js (hunk 3 above)
Comment (si-x-li) [reply backend/routes/courses.js]: done thanks!

File: backend/routes/courses.js (hunk 1 above)
Comment (si-x-li) [reply backend/routes/courses.js]: finished ty!

File: backend/logic/courses/courses.js (hunk 2 above)
Comment (KevenLYF) [reply backend/logic/courses/courses.js]: yes similar to `finally` in Java, it will always execute

**Comment by KevenLYF** [comment]
the commit history is a bit messy, can you do a rebase? Also what do you think about putting `connectoin.release()` in to `finally` block ?

**Comment by si-x-li** [comment]
Git rebase completed and changed to use finally

**Pull request merged**